Added incomplete MACS2.

gunjanbaid opened the pull request:
Added initial code for MACS2 in Cannoli. Currently only added support for `callpeak` function and one parameter, `-t`. 

Things to discuss:
- which other functions and parameters to add support for
- how to support multiple possible input formats
- how to process output, which is four different files

@akmorrow13

gunjanbaid committed:
Added incomplete MACS2.

akmorrow13 commented on src/main/scala/org/bdgenomics/cannoli/Macs2.scala:
```diff
@@ -0,0 +1,80 @@
+  def run(sc: SparkContext) {
+    // not sure if FragmentRDD the right choice here
+    // -Gunjan
+    val input: FragmentRDD = sc.loadFragments(args.inputPath)
```
From looking at MACS2, it like the formatting options are AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE. It is just a matter of which ones we want to support. I think BAM and BED can be easily translated to loadAlignments() and loadFeatures(), respsectively. However, it does not look like you are actually using the input value that is loaded from the RDD. Ideally, we would want to use the RDD of features or alignments as input and pipe it in to MACS2, as seen here with the bowtie implementation: https://github.com/bigdatagenomics/cannoli/blob/master/src/main/scala/org/bdgenomics/cannoli/Bowtie.scala

gunjanbaid replied:
Sounds good, I'll add in support for both BAM and BED files. 

I did not pipe the RDD data into the command yet because I was unsure on how to use the `pipe` method. Using `pipe` requires specifying the output record type. The bowtie example is such:
```
val bowtie2Command = "bowtie2 -x " + args.indexPath + " -U -"
val output: AlignmentRecordRDD = input.pipe[AlignmentRecord, AlignmentRecordRDD, InterleavedFASTQInFormatter](bowtie2Command)
      .transform(_.cache())
```
`AlignmentRecord` and `AlignmentRecordRDD` are specified since Bowtie2 outputs records in SAM format. However, for MACS2, the `callpeak` function does not output any records, but instead creates four files. Is `pipe` the wrong method to be using here?

akmorrow13 replied:
We probably want to find a good way to pipe the final results that are called back into an RDD for the user to access. I guess this wouldn't necessarily call the pipe method.

gunjanbaid committed:
Merge branch 'master' of https://github.com/bigdatagenomics/cannoli into macs2

gunjanbaid committed:
Added script for running macs2.

gunjanbaid commented:
Hi @akmorrow13, I added updates files so far. I was able to work around the initial issues with the pipe API. I used `macs2 callpeak -t /dev/stdin` to have MACS2 read from stdin, instead of another input file. I also added a wrapper script for the `pipe` method in order to run 2 commands (`macs2 callpeak -t /dev/stdin` and `cat NA_summits.bed`). This allows me to output the BED entries to stdout so that I can read them back into an RDD.

The problem I am running into now is that macs2 doesn't seem to run properly since the input is coming from an RDD. I've pasted the normal MACS2 output and also the output I am getting with cannoli. When running with cannoli, MACS2 warns that I have "Too few paired peaks" and does not create the files it should. The output of the command is all jumbled up. I think the error is from each partition being processed separately (?).

Should I be doing some sort of preprocessing to this RDD to make sure everything feeds into MACS2 correctly? One thought I had was to repartition the RDD to have just 1 partition. I couldn't get this to work due to type errors, but it seems like that probably isn't the best way to go in order to make the most of Spark here. Thoughts? Let me know if any of this is unclear, I can clarify more.

**MACS2 output should look kind of like this:**
```
$ macs2 callpeak -t test.bed
INFO  @ Tue, 11 Jul 2017 17:11:58:
# Command line: callpeak -t test.bed
# ARGUMENTS LIST:
# name = NA
# format = AUTO
# ChIP-seq file = ['test.bed']
# control file = None
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff = 5.00e-02
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 10000 bps
# Broad region calling is off
# Paired-End mode is off

INFO  @ Tue, 11 Jul 2017 17:11:58: #1 read tag files...
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 read treatment tags...
INFO  @ Tue, 11 Jul 2017 17:11:58: Detected format is: BED
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 tag size is determined as 100 bps
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 tag size = 100
INFO  @ Tue, 11 Jul 2017 17:11:58: #1  total tags in treatment: 199977
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 user defined the maximum tags...
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Tue, 11 Jul 2017 17:11:58: #1  tags after filtering in treatment: 199583
INFO  @ Tue, 11 Jul 2017 17:11:58: #1  Redundant rate of treatment: 0.00
INFO  @ Tue, 11 Jul 2017 17:11:58: #1 finished!
INFO  @ Tue, 11 Jul 2017 17:11:58: #2 Build Peak Model...
INFO  @ Tue, 11 Jul 2017 17:11:58: #2 looking for paired plus/minus strand peaks...
INFO  @ Tue, 11 Jul 2017 17:11:59: #2 number of paired peaks: 4488
INFO  @ Tue, 11 Jul 2017 17:11:59: start model_add_line...
INFO  @ Tue, 11 Jul 2017 17:11:59: start X-correlation...
INFO  @ Tue, 11 Jul 2017 17:11:59: end of X-cor
INFO  @ Tue, 11 Jul 2017 17:11:59: #2 finished!
INFO  @ Tue, 11 Jul 2017 17:11:59: #2 predicted fragment length is 254 bps
INFO  @ Tue, 11 Jul 2017 17:11:59: #2 alternative fragment length(s) may be 254 bps
INFO  @ Tue, 11 Jul 2017 17:11:59: #2.2 Generate R script for model : NA_model.r
INFO  @ Tue, 11 Jul 2017 17:11:59: #3 Call peaks...
INFO  @ Tue, 11 Jul 2017 17:11:59: #3 Pre-compute pvalue-qvalue table...
INFO  @ Tue, 11 Jul 2017 17:11:59: #3 Call peaks for each chromosome...
INFO  @ Tue, 11 Jul 2017 17:12:00: #4 Write output xls file... NA_peaks.xls
INFO  @ Tue, 11 Jul 2017 17:12:00: #4 Write peak in narrowPeak format file... NA_peaks.narrowPeak
INFO  @ Tue, 11 Jul 2017 17:12:00: #4 Write summits bed file... NA_summits.bed
INFO  @ Tue, 11 Jul 2017 17:12:00: Done!
```

**Here is what some of the output from running cannoli looks like:**
```
$ ADAM_MAIN=org.bdgenomics.cannoli.Cannoli \
    ../adam/bin/adam-submit \
    --jars target/cannoli-spark2_2.11-0.1-SNAPSHOT.jar \
    -- \
    macs2 callpeak -t test.bed
Using ADAM_MAIN=org.bdgenomics.cannoli.Cannoli
Using SPARK_SUBMIT=/usr/local/bin/spark-submit
2017-07-11 17:15:57 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
INFO  @ Tue, 11 Jul 2017 17:16:09:
# Command line: callpeak -t /dev/stdin
# ARGUMENTS LIST:
# name = NA
# format = AUTO
# ChIP-seq file = ['/dev/stdin']
# control file = None
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff = 5.00e-02
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 10000 bps
# Broad region calling is off
# Paired-End mode is off

INFO  @ Tue, 11 Jul 2017 17:16:09:
# Command line: callpeak -t /dev/stdin
# ARGUMENTS LIST:
# name = NA
# format = AUTO
# ChIP-seq file = ['/dev/stdin']
# control file = None
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff = 5.00e-02
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 10000 bps
# Broad region calling is off
# Paired-End mode is off

INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read tag files...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read treatment tags...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read tag files...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read treatment tags...
INFO  @ Tue, 11 Jul 2017 17:16:09:
# Command line: callpeak -t /dev/stdin
# ARGUMENTS LIST:
# name = NA
# format = AUTO
# ChIP-seq file = ['/dev/stdin']
# control file = None
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff = 5.00e-02
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 10000 bps
# Broad region calling is off
# Paired-End mode is off

INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read tag files...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read treatment tags...
INFO  @ Tue, 11 Jul 2017 17:16:09:
# Command line: callpeak -t /dev/stdin
# ARGUMENTS LIST:
# name = NA
# format = AUTO
# ChIP-seq file = ['/dev/stdin']
# control file = None
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff = 5.00e-02
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 10000 bps
# Broad region calling is off
# Paired-End mode is off

INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read tag files...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 read treatment tags...
INFO  @ Tue, 11 Jul 2017 17:16:09: Detected format is: BED
INFO  @ Tue, 11 Jul 2017 17:16:09: Detected format is: BED
INFO  @ Tue, 11 Jul 2017 17:16:09: Detected format is: BED
INFO  @ Tue, 11 Jul 2017 17:16:09: Detected format is: BED
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 tag size is determined as 100 bps
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 tag size = 100
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  total tags in treatment: 7564
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 user defined the maximum tags...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  tags after filtering in treatment: 7549
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 tag size is determined as 100 bps
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  Redundant rate of treatment: 0.00
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 tag size = 100
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 finished!
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  total tags in treatment: 8984
INFO  @ Tue, 11 Jul 2017 17:16:09: #2 Build Peak Model...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 user defined the maximum tags...
INFO  @ Tue, 11 Jul 2017 17:16:09: #2 looking for paired plus/minus strand peaks...
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  tags after filtering in treatment: 8956
INFO  @ Tue, 11 Jul 2017 17:16:09: #1  Redundant rate of treatment: 0.00
INFO  @ Tue, 11 Jul 2017 17:16:09: #1 finished!
INFO  @ Tue, 11 Jul 2017 17:16:09: #2 Build Peak Model...
INFO  @ Tue, 11 Jul 2017 17:16:09: #2 looking for paired plus/minus strand peaks...
INFO  @ Tue, 11 Jul 2017 17:16:09: #2 number of paired peaks: 0
WARNING @ Tue, 11 Jul 2017 17:16:09: Too few paired peaks (0) so I can not build the model! Broader your MFOLD range parameter may erase this error. If it still can't build the model, we suggest to use --nomodel and --extsize 147 or other fixed number instead.
WARNING @ Tue, 11 Jul 2017 17:16:09: Process for pairing-model is terminated!
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 number of paired peaks: 0
WARNING @ Tue, 11 Jul 2017 17:16:10: Too few paired peaks (0) so I can not build the model! Broader your MFOLD range parameter may erase this error. If it still can't build the model, we suggest to use --nomodel and --extsize 147 or other fixed number instead.
WARNING @ Tue, 11 Jul 2017 17:16:10: Process for pairing-model is terminated!
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 tag size is determined as 100 bps
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 tag size = 100
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  total tags in treatment: 8985
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 user defined the maximum tags...
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  tags after filtering in treatment: 8968
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  Redundant rate of treatment: 0.00
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 finished!
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 Build Peak Model...
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 looking for paired plus/minus strand peaks...
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 number of paired peaks: 0
WARNING @ Tue, 11 Jul 2017 17:16:10: Too few paired peaks (0) so I can not build the model! Broader your MFOLD range parameter may erase this error. If it still can't build the model, we suggest to use --nomodel and --extsize 147 or other fixed number instead.
WARNING @ Tue, 11 Jul 2017 17:16:10: Process for pairing-model is terminated!
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 tag size is determined as 100 bps
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 tag size = 100
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  total tags in treatment: 17645
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 user defined the maximum tags...
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  tags after filtering in treatment: 17622
INFO  @ Tue, 11 Jul 2017 17:16:10: #1  Redundant rate of treatment: 0.00
INFO  @ Tue, 11 Jul 2017 17:16:10: #1 finished!
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 Build Peak Model...
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 looking for paired plus/minus strand peaks...
INFO  @ Tue, 11 Jul 2017 17:16:10: #2 number of paired peaks: 0
WARNING @ Tue, 11 Jul 2017 17:16:10: Too few paired peaks (0) so I can not build the model! Broader your MFOLD range parameter may erase this error. If it still can't build the model, we suggest to use --nomodel and --extsize 147 or other fixed number instead.
WARNING @ Tue, 11 Jul 2017 17:16:10: Process for pairing-model is terminated!
INFO  @ Tue, 11 Jul 2017 17:16:11:
```

akmorrow13 commented on src/main/scala/org/bdgenomics/cannoli/Cannoli.scala:
```diff
@@ -30,6 +30,7 @@ object Cannoli {
       Bowtie2,
       Bwa,
       Freebayes,
+      Macs2,
```
MACS2

akmorrow13 commented on src/main/scala/org/bdgenomics/cannoli/Macs2.scala:
```diff
@@ -0,0 +1,81 @@
+    implicit val tFormatter = BEDInFormatter
+    implicit val uFormatter = new BEDOutFormatter
+    val output: FeatureRDD = input.pipe(macs2Command)
+    output.save("output.bed", true, false)
```
OUTPUT parameter instead of output.bed

akmorrow13 commented:
Awesome @gunjanbaid ! From out last conversation, I was under the impression that macs2 would not run distributed, but be able to run in parallel on many different files. I don't think there is a good use case for actually running MACS2 distributed, but running MACS2 instance on many different examples. Therefore, if you are taking a file and partitioning it then running MACS2 on each separate partition, it would not work. In order for something like that to work, you would have to rewrite MACS2, which is not what I think you want to do. I think you would want to partitionby file, then run MACS2 on each partition.

akmorrow13 commented on src/main/scala/org/bdgenomics/cannoli/Macs2.scala:
```diff
@@ -0,0 +1,81 @@
+  def run(sc: SparkContext) {
+    val macs2Command = "./run-macs2.sh"
+
+    val input: FeatureRDD = sc.loadFeatures(args.treatmentFile)
```
I think the more probable use case here is peak calling alignments. What was your reasoning for features?

gunjanbaid replied:
I was planning to add support for both BAM and BED files. I removed the BAM file support from my code temporarily as I was just testing with BED files. But adding that back in now!

gunjanbaid commented:
Ok, I see why partitioning a single file is not what I should be doing here. Could you clarify a bit more on the expected output for peak calling multiple files in parallel? Should I:

1. Input multiple files (A, B, C) and get one set of output files (output) 
2. OR input multiple files (A, B, C) and get multiple sets of output files (outputA, outputB, outputC)

akmorrow13 commented:
I could see you taking many files (A ,B, C) as input. I could see you doing output either way. If you combine all results into 1 file, you would just need a way to identify each peak by the file it originated from, which could probably be fit somewhere in the Features schema (possibly in the source field, found here https://github.com/bigdatagenomics/bdg-formats/blob/master/src/main/resources/avro/bdg.avdl#L1166). I believe this is more 'Sparky' than saving a bunch of separate files.

gunjanbaid committed:
Now creating RDD of filenames, instead of features.

gunjanbaid committed:
Merge branch 'master' of https://github.com/bigdatagenomics/cannoli into macs2

gunjanbaid commented:
So I decided to implement option 2 (input multiple files, output multiple files) because that seemed cleaner. I changed the design a bit: instead of creating an RDD of features/alignments, I just created an RDD of filenames (one filename per partition), which I piped to the macs2 command. This allowed me to create separate files for each input file really easily. So for example, running this command
```
$ ADAM_MAIN=org.bdgenomics.cannoli.Cannoli \
    ../adam/bin/adam-submit \
    --jars target/cannoli-spark2_2.11-0.1-SNAPSHOT.jar \
    -- \
    macs2 callpeak test.bed,test2.bed output
```
will create output files with this structure
```
output/  
    test.bed_output/
        test.bed_model.r
        test.bed_peaks.narrowPeak
        test.bed_peaks.xls
        test.bed_summits.bed 
    test2.bed_output/
        test2.bed_model.r
        test2.bed_peaks.narrowPeak
        test2.bed_peaks.xls
        test2.bed_summits.bed 
```
This looks a lot different from the implementations of other cannoli wrappers. Let me know if this is the wrong way to go, in terms of design.

gunjanbaid commented on src/main/scala/org/bdgenomics/cannoli/Macs2.scala:
```diff
@@ -0,0 +1,84 @@
+    val input = sc.parallelize(inputFiles, inputFiles.length)
+    val output = input.pipe(macs2Command)
+    // force evaluation
+    output.count()
```
I didn't use `output.save()` here because MACS2 will create the output files anyways, so `save()` seemed redundant. Using `count()` as a placeholder just to trigger the evaluation.

gunjanbaid commented on src/main/scala/org/bdgenomics/cannoli/Macs2.scala:
```diff
@@ -0,0 +1,84 @@
+    val inputFiles = args.inputPath.split(",")
+    // create RDD of filenames, one filename per partition
+    val input = sc.parallelize(inputFiles, inputFiles.length)
+    val output = input.pipe(macs2Command)
```
This is not the same is `GenomicRDD.pipe()`. This is using [`RDD.pipe()`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L724) since `input` is an RDD of strings.

gunjanbaid commented:
@fnothaft @heuermh If you have any comments, that would be really helpful. Thanks!

More explanation for current approach and problems with other approaches:

#### My approach
* Input RDD[String] containing filenames, one filename per partition
* Pass input through RDD.pipe() to run-macs2.sh, which runs MACS2 on each filename
* Output RDD[String], where each element of the RDD is a line from BED file created by MACS2
* Don’t do anything with the output of pipe, since MACS2 will create four files. I call output.count() just to trigger the creation of the four files.

#### Other approaches and problems
* Input GenomicRDD to pipe
    * Need to ensure each partition contains entries from a separate file
    * Was not sure how to specify different output filenames for each input file in run-macs2.sh (running MACS2 on partition 1 should create outputfile1, running on partition 2 should create outputfile2, etc.). Output name is specified using --outdir in the MACS2 command.
* Output GenomicRDD to pipe
    * If input is not GenomicRDD, output will be RDD[String]. Cannot convert RDD[String] to GenomicRDD (right?)
    * If output is GenomicRDD, not sure if I can ensure that each partition comes from a separate file
* Output RDD[String] from pipe, each element is the filename created
    * No way to know which node the output files were created on, so can’t read them back into a GenomicRDD using a path

gunjanbaid committed:
Merge branch 'master' into macs2

gunjanbaid committed:
Able to write, read, delete MACS2 output files.

gunjanbaid committed:
Merge branch 'master' of https://github.com/bigdatagenomics/cannoli into macs2
